src/hooks/login/useWalletConnectV2Login.ts

```diff
@@ -0,0 +1,243 @@
+  const isLoading = !hasWcUri;
+  const uriDeepLink = hasWcUri
+    ? `${walletConnectDeepLink}?wallet-connect=${encodeURIComponent(wcUri)}`
+    : null;
+
+  useUpdateEffect(() => {
+    generateWcUri();
+  }, [token]);
+
+  useUpdateEffect(() => {
+    providerRef.current = provider;
+  }, [provider]);
+
+  const handleOnLogout = () => {
+    logout(logoutRoute);
+  };
+
+  async function handleOnLogin() {
+    try {
+      const provider = providerRef.current;
+      if (isLoggedIn) {
+        return;
+      }
+      if (provider == null) {
+        return;
+      }
+      if (Object.keys(provider).length === 0) {
+        return;
+      }
+      const address = await provider.getAddress();
+      const signature = await provider.getSignature();
+      const hasSignature = Boolean(signature);
+      const loginActionData = {
+        address: address,
+        loginMethod: LoginMethodsEnum.walletconnectv2
+      };
+
+      const loginData = {
+        logoutRoute: logoutRoute,
+        loginType: 'walletconnectv2',
+        callbackRoute: callbackRoute ?? window.location.href
+      };
+
+      if (hasSignature) {
+        dispatch(setWalletConnectLogin(loginData));
+        dispatch(setTokenLoginSignature(signature));
+      } else {
+        dispatch(setWalletConnectLogin(loginData));
+      }
+
+      dispatch(loginAction(loginActionData));
+      optionalRedirect(callbackRoute, onLoginRedirect);
+    } catch (err) {
+      setError(WalletConnectV2Error.invalidAddress);
+      console.error(err);
+    }
+  }
+
+  async function initiateLogin(loginProvider = true) {
+    const shouldGenerateWcUri = loginProvider && !wcUri;
+    if (
+      !walletConnectV2ProjectId ||
+      !walletConnectV2RelayAddress ||
+      (providerRef?.current?.isInitialized?.() && !shouldGenerateWcUri)
+    ) {
+      return;
+    }
+
+    const providerHandlers = {
+      onClientLogin: handleOnLogin,
+      onClientLogout: handleOnLogout
+    };
+
+    const newProvider = new WalletConnectProviderV2(
+      providerHandlers,
+      chainId,
+      walletConnectV2RelayAddress,
+      walletConnectV2ProjectId
+    );
+
+    await newProvider.init();
+    setAccountProvider(newProvider);
+    setWcPairings(newProvider.pairings);
+    providerRef.current = newProvider;
+    if (loginProvider) {
+      generateWcUri();
+    }
+  }
+
+  async function connectExisting(pairing: PairingTypes.Struct | undefined) {
+    if (!walletConnectV2RelayAddress || !walletConnectV2ProjectId) {
+      setError(WalletConnectV2Error.invalidConfig);
+      return;
+    }
+    if (!pairing || !pairing.topic) {
+      return;
+    }
+
+    try {
+      const { approval } = await providerRef.current?.connect({
+        topic: pairing.topic
+      });
+      if (token) {
+        dispatch(setTokenLogin({ loginToken: token }));
+      }
+
+      try {
+        await providerRef.current?.login({ approval });
+      } catch {
+        await initiateLogin();
+      }
+    } catch {
```

Comment:
Would be very nice to log the original error before setting a custom error. It will help a lot on debugging